server/src/main/java/org/opensearch/index/ShardIndexingPressureMemoryManager.java

```diff
@@ -0,0 +1,458 @@
+        return totalLastSuccessfulRequestLimitsBreachedRejections.get();
+    }
+
+    long getTotalThroughputDegradationLimitsBreachedRejections() {
+        return totalThroughputDegradationLimitsBreachedRejections.get();
+    }
+
+    /**
+     * Verifies and returns true if the shard limit is hard-breached i.e. shard limit cannot be increased further. Otherwise
+     * increases the shard limit and returns false.
+     */
+    private boolean onShardLimitBreached(long nodeTotalBytes, long nodeLimit, long requestStartTime, OperationTracker operationTracker,
+                                         BooleanSupplier increaseShardLimitSupplier) {
+
+        // Secondary Parameters (i.e. LastSuccessfulRequestDuration and Throughput) is taken into consideration when
+        // the current node utilization is greater than primary_parameter.node.soft_limit of total node limits.
+        if(((double)nodeTotalBytes / nodeLimit) < this.nodeSoftLimit) {
+            boolean isShardLimitsIncreased = increaseShardLimitSupplier.getAsBoolean();
+            if (isShardLimitsIncreased == false) {
+                operationTracker.getRejectionTracker().incrementNodeLimitsBreachedRejections();
+                this.totalNodeLimitsBreachedRejections.incrementAndGet();
+            }
+            return !isShardLimitsIncreased;
+        } else {
+            boolean shardLastSuccessfulRequestDurationLimitsBreached =
+                evaluateLastSuccessfulRequestDurationLimitsBreached(operationTracker.getPerformanceTracker(), requestStartTime);
+
+            if (shardLastSuccessfulRequestDurationLimitsBreached) {
+                operationTracker.getRejectionTracker().incrementLastSuccessfulRequestLimitsBreachedRejections();
+                this.totalLastSuccessfulRequestLimitsBreachedRejections.incrementAndGet();
+                return true;
+            }
+
+            boolean shardThroughputDegradationLimitsBreached =
+                evaluateThroughputDegradationLimitsBreached(operationTracker.getPerformanceTracker(),
+                    operationTracker.getStatsTracker(), primaryAndCoordinatingThroughputDegradationLimits);
+
+            if (shardThroughputDegradationLimitsBreached) {
+                operationTracker.getRejectionTracker().incrementThroughputDegradationLimitsBreachedRejections();
+                this.totalThroughputDegradationLimitsBreachedRejections.incrementAndGet();
+                return true;
+            }
+
+            boolean isShardLimitsIncreased = increaseShardLimitSupplier.getAsBoolean();
+            if (isShardLimitsIncreased == false) {
+                operationTracker.getRejectionTracker().incrementNodeLimitsBreachedRejections();
+                this.totalNodeLimitsBreachedRejections.incrementAndGet();
+            }
+            return !isShardLimitsIncreased;
+        }
+    }
+
+    private boolean increaseShardLimits(ShardId shardId, long nodeLimit,
```

Comment:
Lets add tests for this.

Reply:
All the paths for this are already covered in the tests for the public contracts of the class such as ``isCoordinatingShardLimitBreached``, for ex tests with suffixes:

```
ShardLimitsNotBreached
ShardLimitsIncreasedAndSoftLimitNotBreached
SoftLimitNotBreachedAndNodeLevelRejections
SoftLimitBreachedAndNodeLevelRejections
SoftLimitBreachedAndNoSecondaryParameterBreachedAndNodeLevelRejections